src/Build/Logging/ConsoleLogger.cs

```diff
@@ -72,9 +73,10 @@ public ConsoleLogger()
         /// </summary>
         /// <param name="verbosity">Verbosity level.</param>
         public ConsoleLogger(LoggerVerbosity verbosity) :
-            this(verbosity, Console.Out.Write, BaseConsoleLogger.SetColor, BaseConsoleLogger.ResetColor)
+            this(verbosity, BaseConsoleLogger.Write, BaseConsoleLogger.SetColor, BaseConsoleLogger.ResetColor)
         {
-            // do nothing
+            // set output encoding to UTF-8 to support all languages.
+            Console.OutputEncoding = Encoding.UTF8;
```

Comment:
Should we make this configurable? I'm wondering if there can be any users relying on the current behavior - encoding preserving the system wide setting

Reply:
This seems scary - @nagilson did something with encoding recently to mimic the .NET SDK's behavior. @nagilson can you link to our decision-making process for that change?

Reply:
I think this is a good change but it needs to be made with deep care. Some consoles will react differently to this change and sometimes you need to change the input encoding. There should definitely be an escape hatch if this console is an existing product as it will most likely break people. Also, some versions of Windows 10 do not even support this, so you should probably guard against that.

Reply:
Please see my comments here: https://github.com/dotnet/msbuild/pull/8503

https://learn.microsoft.com/en-us/dotnet/core/compatibility/sdk/8.0/console-encoding this document is probably more digestible but lacks as much technical detail

Reply:
Other option is that user was on a [Windows build that didn't yet support the UTF-8](https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page#set-a-process-code-page-to-utf-8).

Would there be more user bringing up the same issue, we might want to consider adding low-prio message into log when [detecting the case](https://github.com/dotnet/msbuild/blob/42383a83b2b0d1e551cbdb713f9ae3dda27dd04d/src/Framework/EncodingUtilities.cs#L301).

Reply:
> @GangWang01 - can you please try if you can repro the case with .NET 8 RTM (or any .NET 9)? If yes - then we should first investigate why the fix by @nagilson doesn't cover the case.

With .NET 8 (8.0.201) or .NET 9 preview 1, I still could reproduce this issue. The fix by @nagilson works, but requires setting the environment variable `DOTNET_CLI_UI_LANGUAGE`. The issue didn't do this setting.
![net8](https://github.com/dotnet/msbuild/assets/2950449/cc12863d-d4d9-424b-b126-508491c1b8e4)

Reply:
@baronfel - should the msbuild follow the suit of sdk and still require the `DOTNET_CLI_UI_LANGUAGE` here?

Reply:
It would be best to have product parity IMO, but my interest is how the language was different without that variable being set. Was it set in the OS and if so how? (There are several ways to set it and sometimes they conflict with one another.) The same logic should probably trigger in both layers if theres some case where the OS setting is not currently being detected by us, but the ENV var should take precedence.

Reply:
@JanKrivanek I kind of feel that the SDK should use its knowledge of DOTNET_CLI_UI_LANGUAGE to set whatever MSBuild's knob for language is. The SDK is the point at which people interact with MSBuild and it already has a number of ways it customizes the invocation/behavior of MSBuild - this would just be one more.

Reply:
So looking into this bit more closely, I *believe*:

* SDK does the proper thing by checking only the env vars (`DOTNET_CLI_UI_LANGUAGE`, `VSLANG`): https://github.com/dotnet/sdk/blob/5c966a7d1acd6fe5c27f09b81a7652f76fd0ac23/src/Cli/Microsoft.DotNet.Cli.Utils/UILanguageOverride.cs#L52-L84 and specifically not checking [System Locale](https://learn.microsoft.com/en-us/windows-hardware/customize/desktop/unattend/microsoft-windows-international-core-winpe-systemlocale) or [Display Language](https://support.microsoft.com/en-us/windows/manage-the-input-and-display-language-settings-in-windows-12a10cb4-8626-9b77-0ccb-5013e0c7c7a2), as those should be automatically reflected in  `CultureInfo.CurrentUICulture`
* MSBuild almost has a feature parity with the SDK (with the Noah change #8503), it just misse the setting of the encoding if needed by locale
* Gangs changes here almost fixes the missing part - it is just too wide - and needs to be conditioned, [similarly as in sdk](https://github.com/dotnet/sdk/blob/5c966a7d1acd6fe5c27f09b81a7652f76fd0ac23/src/Cli/Microsoft.DotNet.Cli.Utils/UILanguageOverride.cs#L27-L35)
* The env vars checking logic should probably still stay in MSBuild, even if it feels duplicated with sdk - as msbuild can be invoked from VS or directly and since that's currently the main way how sdk communicates locale to other child processes anyways.

All that being - if you do not disagree @baronfel and @nagilson - we should probably keep the @GangWang01 changes, just move them to front of [`SetConsoleUI`](https://github.com/dotnet/msbuild/blob/5ba4508408b1c38aa95dcf62dc237b9a23f2523b/src/MSBuild/XMake.cs#L1834), condition them for non-english locales on OSes that support UTF8 (same as sdk does) and ensure that it plays well together with [the current codepage checking logic](https://github.com/dotnet/msbuild/blob/5ba4508408b1c38aa95dcf62dc237b9a23f2523b/src/MSBuild/XMake.cs#L1861-L1871).

Reply:
We might also need to consider another case that is to output non-English(Korean/Japanese/Chinese) message with msbuild `Message`/`Warning` task in English system locale.

Reply:
> We might also need to consider another case that is to output non-English(Korean/Japanese/Chinese) message with msbuild `Message`/`Warning` task in English system locale.

I believe that from the MSBuild point of view the scenario is still same: observing the `DOTNET_CLI_UI_LANGUAGE`.

If someone wants to have localized tasks on en-US OS, then I believe demanding explicit `DOTNET_CLI_UI_LANGUAGE` set by user is a reasonable requirement here.

Or did you have a different case in mind?

Reply:
The case that often bites us is:

* OS configured for CP 437(really any non-Unicode codepage)
* Project is in path named with Unicode characters that aren't representable in that codepage
* Messages (reasonably!) want to include paths

Reply:
Good point!
Discussed offline with @rainersigwald and we should:
* use UTF-8 (if supported by OS and Console.Out present) unconditionally on Core
* still require explicit opt-in of FullFW (again still checking same conditions)